File: HelloIroh/IrohPeer.swift
import Foundation
import Observation
import IrohLib
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

@MainActor
@Observable
final class IrohPeer {
    enum ConnectionState: Equatable {
        case idle
        case binding
        case ready
        case connecting
        case connected(peerShortId: String)
        case error(String)
    }

    enum TelemetryState: Equatable {
        case off
        case starting
        case active(name: String)
        case error(String)
    }

    let motion: MotionSource
    var endpointId: String = ""
    var state: ConnectionState = .idle
    var remotePosition: SIMD2<Float>? = nil
    var telemetry: TelemetryState = .off
    var apiSecret: String = UserDefaults.standard.string(forKey: IrohPeer.apiSecretKey) ?? ""

    static let apiSecretKey = "iroh.helloiroh.apiSecret"

    private var endpoint: Endpoint?
    private var acceptTask: Task<Void, Never>?
    private var currentSession: PeerSession?
    private var identity: IdentityStore?
    private var services: ServicesClient?

    init(motion: MotionSource) {
        self.motion = motion
    }

    func start() async {
        guard endpoint == nil else { return }
        state = .binding
        let identity = IdentityStore.loadOrCreate()
        self.identity = identity
        endpointId = identity.endpointId
        do {
            let ep = try await Endpoint.bind(options: EndpointOptions(
                preset: presetN0(),
                secretKey: identity.secretKey.toBytes(),
                alpns: [WireFormat.alpn]
            ))
            endpoint = ep
            state = .ready
            acceptTask = Task { [weak self] in
                await self?.runAcceptLoop(ep)
            }
            await startServicesClient()
        } catch {
            state = .error("bind failed: \(error)")
        }
    }

    func saveApiSecret(_ secret: String) async {
        let trimmed = secret.trimmingCharacters(in: .whitespacesAndNewlines)
        apiSecret = trimmed
        UserDefaults.standard.set(trimmed, forKey: Self.apiSecretKey)
        await startServicesClient()
    }

    private func startServicesClient() async {
        services = nil
        let secret = apiSecret
        guard !secret.isEmpty else {
            telemetry = .off
            return
        }
        guard let ep = endpoint else { return }
        telemetry = .starting
        let name = deviceName()
        do {
            let client = try await ServicesClient.create(
                endpoint: ep,
                options: ServicesOptions(apiSecret: secret, name: name)
            )
            services = client
            telemetry = .active(name: name)
        } catch {
            telemetry = .error("\(error)")
        }
    }

    private func deviceName() -> String {
        let short = String(endpointId.prefix(8))
        #if os(iOS)
        return "ios-\(short)"
        #elseif os(macOS)
        return "macos-\(short)"
        #else
        return "ball-\(short)"
        #endif
    }

    func connect(toEndpointIdHex hex: String) async {
        guard let ep = endpoint else { return }
        let trimmed = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let parsed = try? EndpointId.fromString(s: trimmed) else {
            state = .error("invalid endpoint id")
            return
        }
        state = .connecting
        let addr = EndpointAddr(id: parsed, relayUrl: nil, addresses: [])
        do {
            let conn = try await ep.connect(addr: addr, alpn: WireFormat.alpn)
            let bi = try await conn.openBi()
            adoptSession(bi: bi, remoteIdHex: parsed.description)
        } catch {
            state = .error("connect failed: \(error)")
        }
    }

    func copyEndpointIdToClipboard() {
        #if canImport(UIKit)
        UIPasteboard.general.string = endpointId
        #elseif canImport(AppKit)
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(endpointId, forType: .string)
        #endif
    }

    private func runAcceptLoop(_ ep: Endpoint) async {
        while !Task.isCancelled {
            guard let incoming = await ep.acceptNext() else { return }
            do {
                let accepting = try await incoming.accept()
                let alpn = try await accepting.alpn()
                guard alpn == WireFormat.alpn else { continue }
                let conn = try await accepting.connect()
                let bi = try await conn.acceptBi()
                let remoteId = conn.remoteId().description
                adoptSession(bi: bi, remoteIdHex: remoteId)
            } catch {
                continue
            }
        }
    }

    private func adoptSession(bi: BiStream, remoteIdHex: String) {
        currentSession?.stop()
        let session = PeerSession(
            bi: bi,
            getLocalPosition: { [weak self] in
                await MainActor.run { self?.motion.position ?? .zero }
            },
            onRemotePosition: { [weak self] pos in
                Task { @MainActor in self?.remotePosition = pos }
            },
            onClosed: { [weak self] in
                Task { @MainActor in self?.handleSessionClosed() }
            }
        )
        currentSession = session
        session.start()
        let short = String(remoteIdHex.prefix(10))
        state = .connected(peerShortId: short)
    }

    private func handleSessionClosed() {
        guard currentSession != nil else { return }
        currentSession = nil
        remotePosition = nil
        if case .connected = state { state = .ready }
    }
}

File: kotlin-android/app/src/main/java/computer/iroh/pong/net/IrohPeer.kt
package computer.iroh.pong.net

import android.os.Build
import android.util.Log
import computer.iroh.BiStream
import computer.iroh.Endpoint
import computer.iroh.EndpointAddr
import computer.iroh.EndpointId
import computer.iroh.EndpointOptions
import computer.iroh.ServicesClient
import computer.iroh.ServicesOptions
import computer.iroh.presetN0
import computer.iroh.pong.game.MotionSource
import computer.iroh.pong.game.PongGame
import computer.iroh.pong.identity.IdentityStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

private const val TAG = "IrohPeer"

class IrohPeer(
    private val scope: CoroutineScope,
    private val motion: MotionSource,
    private val identity: IdentityStore,
) {
    sealed interface State {
        data object Idle : State
        data object Binding : State
        data object Ready : State
        data object Connecting : State
        data class Connected(val peerShortId: String) : State
        data class Error(val message: String) : State
    }

    sealed interface TelemetryState {
        data object Off : TelemetryState
        data object Starting : TelemetryState
        data class Active(val name: String) : TelemetryState
        data class Error(val message: String) : TelemetryState
    }

    val game = PongGame()

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state.asStateFlow()

    private val _endpointId = MutableStateFlow(identity.endpointId)
    val endpointId: StateFlow<String> = _endpointId.asStateFlow()

    private val _telemetry = MutableStateFlow<TelemetryState>(TelemetryState.Off)
    val telemetry: StateFlow<TelemetryState> = _telemetry.asStateFlow()

    private val _apiSecret = MutableStateFlow(identity.apiSecret)
    val apiSecret: StateFlow<String> = _apiSecret.asStateFlow()

    val isUsingDefaultApiSecret: Boolean get() = _apiSecret.value.isEmpty()

    private var endpoint: Endpoint? = null
    private var acceptJob: Job? = null
    private var session: PeerSession? = null
    private var services: ServicesClient? = null

    fun start() {
        if (endpoint != null) return
        _state.value = State.Binding
        scope.launch {
            try {
                val ep = Endpoint.bind(
                    EndpointOptions(
                        preset = presetN0(),
                        secretKey = identity.secretKey.toBytes(),
                        alpns = listOf(WireFormat.ALPN),
                    ),
                )
                endpoint = ep
                _state.value = State.Ready
                acceptJob = scope.launch { runAcceptLoop(ep) }
                startServicesClient()
            } catch (t: Throwable) {
                _state.value = State.Error("bind failed: ${t.message ?: t}")
            }
        }
    }

    fun connect(toEndpointIdHex: String) {
        val ep = endpoint ?: return
        val trimmed = toEndpointIdHex.trim()
        scope.launch {
            val parsed = try {
                EndpointId.fromString(trimmed)
            } catch (t: Throwable) {
                _state.value = State.Error("invalid endpoint id")
                return@launch
            }
            _state.value = State.Connecting
            try {
                val addr = EndpointAddr(parsed, null, emptyList())
                val conn = ep.connect(addr, WireFormat.ALPN)
                val bi = conn.openBi()
                adoptSession(bi, parsed.toString(), asAuthority = true)
            } catch (t: Throwable) {
                _state.value = State.Error("connect failed: ${t.message ?: t}")
            }
        }
    }

    fun saveApiSecret(secret: String) {
        val trimmed = secret.trim()
        identity.apiSecret = trimmed
        _apiSecret.value = trimmed
        scope.launch { startServicesClient() }
    }

    private suspend fun startServicesClient() {
        services = null
        val ep = endpoint ?: return
        val secret = _apiSecret.value.ifEmpty { DEFAULT_API_SECRET }
        val name = deviceName()
        _telemetry.value = TelemetryState.Starting
        try {
            val client = ServicesClient.create(
                ep,
                ServicesOptions(apiSecret = secret, name = name),
            )
            services = client
            _telemetry.value = TelemetryState.Active(name)
        } catch (t: Throwable) {
            _telemetry.value = TelemetryState.Error("${t.message ?: t}")
        }
    }

    private fun deviceName(): String {
        val short = _endpointId.value.take(8)
        val model = Build.MODEL.lowercase().replace(" ", "-")
        return "android-$model-$short"
    }

    private suspend fun runAcceptLoop(ep: Endpoint) {
        while (currentCoroutineContext().isActive) {
            val incoming = try {
                ep.acceptNext() ?: return
            } catch (t: Throwable) {
                Log.w(TAG, "acceptNext threw", t)
                continue
            }
            try {
                val accepting = incoming.accept()
                val alpn = accepting.alpn()
                if (!alpn.contentEquals(WireFormat.ALPN)) continue
                val conn = accepting.connect()
                val bi = conn.acceptBi()
                val remoteIdHex = conn.remoteId().toString()
                adoptSession(bi, remoteIdHex, asAuthority = false)
            } catch (t: Throwable) {
                Log.w(TAG, "accept flow threw", t)
            }
        }
    }

    private fun adoptSession(bi: BiStream, remoteIdHex: String, asAuthority: Boolean) {
        session?.stop()
        game.resetForNewSession(asAuthority)
        val g = game
        val m = motion
        lateinit var s: PeerSession
        s = PeerSession(
            bi = bi,
            produceFrames = { g.produceTickFrames(m.paddleX) },
            onPaddleReceived = { x -> g.receivedOpponentPaddle(x) },
            onBallReceived = { p -> g.receivedBall(p) },
            // Capture the session identity in the closure so a late-firing
            // onClosed from a previous session can't tear down the current one.
            onClosed = { if (session === s) handleSessionClosed() },
        )
        session = s
        s.start(scope)
        val short = remoteIdHex.take(10)
        _state.value = State.Connected(short)
        Log.d(TAG, "session adopted with $remoteIdHex (authority=$asAuthority)")
    }

    private fun handleSessionClosed() {
        session = null
        game.sessionEnded()
        if (_state.value is State.Connected) _state.value = State.Ready
    }

    companion object {
        const val DEFAULT_API_SECRET = "servicesaaqg6nnf7kr3uiacviqgbxeqconvhuz4ldr5dem4gqhsp3cyat6qxexoctwjsi7m6dh2t2qvfu2yhdoaav6eibaj4aaavhonlixbohceu4aa"
    }
}
